File: src/main/java/io/github/intellij/dlanguage/refactoring/DRefactoringSupportProvider.java
package io.github.intellij.dlanguage.refactoring;

import com.intellij.lang.refactoring.RefactoringSupportProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNamedElement;
import io.github.intellij.dlanguage.psi.DlangIdentifier;
import io.github.intellij.dlanguage.resolve.DResolveUtil;
import org.jetbrains.annotations.NotNull;

import java.util.Set;

/**
 * Created by francis on 4/18/2017.
 */
public class DRefactoringSupportProvider extends RefactoringSupportProvider {
    @Override
    public boolean isSafeDeleteAvailable(@NotNull final PsiElement element) {
        if (!(element instanceof DlangIdentifier))
            return false;
        final Set<PsiNamedElement> resolve = DResolveUtil.Companion.getInstance(element.getProject()).findDefinitionNode((PsiNamedElement) element, false);
        return resolve.size() == 1;
    }


}

File: src/main/kotlin/io/github/intellij/dlanguage/refactoring/DRefactoringSupportProvider.kt
package io.github.intellij.dlanguage.refactoring

import com.intellij.lang.refactoring.RefactoringSupportProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiPolyVariantReference
import io.github.intellij.dlanguage.psi.interfaces.DNamedElement

/**
 * Created by francis on 4/18/2017.
 */
class DRefactoringSupportProvider : RefactoringSupportProvider() {
    override fun isSafeDeleteAvailable(element: PsiElement): Boolean {
        if (element !is DNamedElement) return false
        val reference = element.reference
        if (reference is PsiPolyVariantReference) {
            return reference.multiResolve(false).isEmpty()
        }
        return reference?.resolve() == null
    }
}
